src/main/java/seedu/address/logic/commands/room/EditRoomCommand.java

```diff
@@ -0,0 +1,220 @@
+        Index index = model.getModifiableRoomList().checkIfRoomPresent(roomNumberToEdit);
+
+        if (index.getZeroBased() == 0) {
+            throw new CommandException(MESSAGE_INVALID_ROOM_NOT_FOUND);
+        }
+
+        Room roomToEdit = lastShownList.get(index.getZeroBased() - 1);
+        Room editedRoom = createEditedRoom(model, roomToEdit, editRoomDescriptor);
+
+        if (!roomToEdit.isSameRoom(editedRoom) && model.containsRoom(editedRoom)) {
+            throw new CommandException(MESSAGE_DUPLICATE_ROOM);
+        }
+
+        model.setSingleRoom(roomToEdit, editedRoom);
+        return new CommandResult(String.format(MESSAGE_EDIT_ROOM_SUCCESS, editedRoom));
+    }
+
+    /**
+     * Creates and returns a {@code Room} with the details of {@code roomToEdit}
+     * edited with {@code editRoomDescriptor}.
+     *
+     * @param model Current model
+     * @param roomToEdit Room that is to be edited.
+     * @param editRoomDescriptor Details to edit the room with.
+     * @return Room that has been edited.
+     */
+    private static Room createEditedRoom(Model model, Room roomToEdit,
+                                         EditRoomDescriptor editRoomDescriptor) throws CommandException {
+        assert (roomToEdit != null);
+        int updatedRoomNumber = editRoomDescriptor.getRoomNumber().orElse(roomToEdit.getRoomNumber());
+        boolean clearRoom = (editRoomDescriptor.getIsOccupied().orElse(null) != null); //non-null just means clear room
+        boolean hasNewPatient = (editRoomDescriptor.getPatientName().orElse(null) != null);
+        boolean alreadyOccupied = roomToEdit.isOccupied();
+
+        if (clearRoom && !hasNewPatient) {
+            //case 1: clear room
+            return new Room(updatedRoomNumber);
+        } else if (!clearRoom && !hasNewPatient && alreadyOccupied) {
+            //case 2: change room number in a room with patient already
+            return new Room(updatedRoomNumber, roomToEdit.getPatient());
+        } else if (!clearRoom && !hasNewPatient) {
+            //case 3: change room number in an empty room
+            return new Room(updatedRoomNumber);
+        }
+
+        //case 4: allocate patient to the room
+        Name patientName = editRoomDescriptor.getPatientName().get(); //definitely has name
+        List<Patient> patientList = model.getFilteredPatientList();
+        Room updatedRoom = isValidPatient(patientList, patientName, updatedRoomNumber);
+        return updatedRoom;
+    }
+
+    /**
+     * Checks if the patient is valid and exists in the application.
+     *
+     * @param patientList List of the patients in the application.
+     * @param patientName Name of the patient.
+     * @param updatedRoomNumber Room number of room.
+     * @return Room that is updated with the new patient and room number.
+     */
+    private static Room isValidPatient(List<Patient> patientList, Name patientName,
```

Comment:
I think this method should be changed with a name getValidpatient and should return a patient as the name suggests. isValidPatient is like a boolean method and should only return a boolean

Reply:
okayy, i will change it later! thanks for pointing it out